### ISSUE TITLE
kaiming.conf for ImageNet, NaN problems

### ISSUE THREAD

**Issue by zhongwen** [dmlc/cxxnet#123]
I met NaN (I added a NaN detect [here](https://github.com/dmlc/cxxnet/blob/master/src/utils/metric.h#L126) as [below](https://github.com/dmlc/cxxnet/blob/master/src/utils/metric.h#L128) when I went after round 6 (at least twice from the start)

So I delete some models and deleted the added NaN detect assert, continue to train the networks. Here are my results.

round        4:[   10010] 3433 sec elapsed[5]   train-rec@1:0.169138    train-rec@5:0.374235    val-rec@1:0.20662       val-rec@5:0.43274
round        5:[   10010] 6866 sec elapsed[6]   train-rec@1:0.178102    train-rec@5:0.387523    val-rec@1:0.19768       val-rec@5:0.41596
round        6:[   10010] 10303 sec elapsed[7]  train-rec@1:0.174113    train-rec@5:0.378702    val-rec@1:0.22226       val-rec@5:0.45216
round        7:[   10010] 13746 sec elapsed[8]  train-rec@1:0.0391093   train-rec@5:0.0873142   val-rec@1:0.001 val-rec@5:0.005
round        8:[   10010] 17182 sec elapsed[9]  train-rec@1:0.000868663 train-rec@5:0.00461804  val-rec@1:0.001 val-rec@5:0.005
round        9:[   10010] 20627 sec elapsed[10] train-rec@1:0.000923295 train-rec@5:0.00454389  val-rec@1:0.001 val-rec@5:0.005
round       10:[   10010] 24070 sec elapsed[11] train-rec@1:0.00091393  train-rec@5:0.00453843  val-rec@1:0.001 val-rec@5:0.005
round       11:[   10010] 27505 sec elapsed[12] train-rec@1:0.000916271 train-rec@5:0.00453141  val-rec@1:0.001 val-rec@5:0.005

So things still went wrong after round 7, rec@1 and rec@5 are random.

Another problem is that when I chang the initialization from xavier to kaiming, NaN occurs in the very beginning (batch 2 in round 0).

Could you please help me on this? Thanks.

**Comment by antinucleon** [comment]
What is your GPU Model? Recently I trained 4 kaiming net. On Titan X and 2 980 GTX are fine. On 780 + 980 SLI machine, I met same problem like you. My solution is deleted abnormal model then continue. Finally I can get model with reasonable result.

**Comment by antinucleon** [comment]
For kaiming init, please read his original paper:not all layers can be filled by the formula. Better way is manually set init method/range.

**Comment by zhongwen** [comment]
I have been using 4 K20m. 
I have deleted some abnormal models but still met this problem in the following training process. 

For kaiming init, I changed the fc layers into the settings of his paper. It works. Thanks.

**Comment by antinucleon** [comment]
Yes. On 780+980 machine, I met this problem about every 7-10 round. you need to delete both abnormal model and previous one or two models. if possible, could you turn on ECC to help us check whether it is caused by memory error? Thanks

**Comment by zhongwen** [comment]
ECC for all the cards have been turned on, I do not have privilege to turn them off.

**Comment by winstywang** [comment]
I have confirmed with one of my friend who uses four K40 training on this net. This config just works well for him. It is a bit wired in this case that we cannot diagnose which part goes wrong based on the information you provided. Could you try to start from a lower learning rate, say 0.005?

**Comment by tqchen** [comment]
If the problem was overflowing momentum or gradient, I have made a gradient clipping option in SGD solver which was by default turned off https://github.com/dmlc/cxxnet/blob/master/src/updater/param.h#L55

**Comment by zhongwen** [comment]
@winstywang I put another two cards to run on lr = 0.005, it can run into round 8, though *at the first two rounds, it get val-rec@1 as random*, and it has much lower performance than the net with lr = 0.01 at the same round (as expected).

@tqchen Thanks for pointing out this. What would you suggest for a good clip_gradient value?

**Comment by zhongwen** [comment]
I set clip_gradient = 1, then the net can run smoothly.

**Comment by tqchen** [comment]
great!  I am glad that it works our for you!

**Comment by zhongwen** [comment]
b.t.w. It should be for NAG as well?
